include/rcutils/testing/fault_injection.h

```diff
@@ -0,0 +1,128 @@
+ * to an incrementing count variable.
+ *
+ *  for (int i = 0; i < SUFFICIENTLY_LARGE_ITERATION_COUNT; ++i) {
+ *    RCUTILS_FAULT_INJECTION_SET_COUNT(i);
+ *    ... // Call function under test
+ *  }
+ * ASSERT_LT(RCUTILS_FAULT_INJECTION_NEVER_FAIL, RCUTILS_FAULT_INJECTION_GET_COUNT());
+ *
+ * Where SUFFICIENTLY_LARGE_ITERATION_COUNT is a value larger than the maximum expected calls to
+ * `RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR`. This last assertion just insures that your choice for
+ * SUFFICIENTLY_LARGE_ITERATION_COUNT was large enough. To avoid having to choose this count
+ * yourself, you can use a do-while loop.
+ *
+ * int i = 0;
+ * do {
+ *   RCUTILS_FAULT_INJECTION_SET_COUNT(i++);
+ *    ... // Call function under test
+ * } while (RCUTILS_FAULT_INJECTION_GET_COUNT() <= RCUTILS_FAULT_INJECTION_NEVER_FAIL);
+ *
+ * \param count The count to set the fault injection counter to. If count is negative, then fault
+ * injection errors will be disabled. The counter is globally initialized to
+ * RCUTILS_FAULT_INJECTION_NEVER_FAIL.
+ */
+RCUTILS_PUBLIC
+void
+rcutils_fault_injection_set_count(int count);
+
+/**
+ * \def RCUTILS_FAULT_INJECTION_GET_COUNT
+ * \brief Atomically get the fault injection counter value
+ *
+ * Use this macro after running the code under test to check whether the counter reached a negative
+ * value. This is helpful so you can verify that you ran the fault injection test in a loop a
+ * sufficient number of times. Likewise, if the code under test returned with an error, but the
+ * count value was greater or equal to 0, then the failure was not caused by the fault injection
+ * counter.
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+int_least64_t
+rcutils_fault_injection_get_count(void);
+
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+int_least64_t
+_rcutils_fault_injection_maybe_fail(void);
+
+/**
+ * \def RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR
+ * \brief This macro checks and decrements a static global variable atomic counter and returns
+ * `return_value_on_error` if 0.
+ *
+ * This macro is not a function itself, so when this macro returns it will cause the calling
+ * function to return with the return value.
+ *
+ * Set the counter with `RCUTILS_FAULT_INJECTION_SET_COUNT`. If the count is less than 0, then
+ * `RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR` will not cause an early return.
+ *
+ * This macro is thread-safe, and ensures that at most one invocation results in a failure for each
+ * time the fault injection counter is set with `RCUTILS_FAULT_INJECTION_SET_COUNT`
+ *
+ * \param return_value_on_error the value to return in the case of fault injected failure.
+ */
+#define RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR(return_value_on_error) \
+  if (RCUTILS_FAULT_INJECTION_FAIL_NOW == _rcutils_fault_injection_maybe_fail()) { \
+    printf( \
+      "%s:%d Injecting fault and returning " #return_value_on_error "\n", __FILE__, __LINE__); \
```

Comment:
@brawner why not using debug `logs` or safe fwrites to `stderr`?

Reply:
> RCUTILS_LOG_DEBUG is not defined inside the rcutils library unfortunately.

Just saw this. That's unfortunate, yeah.

Reply:
Hmm wait a sec, `logging_macros.h` is available at build time (see [CMakeLists.txt](https://github.com/ros2/rcutils/blob/ed1f549b02a2054597aa339cbf529ab1e2151baf/CMakeLists.txt#L111-L120)).

Reply:
I'd bet that's because the library doesn't list the custom command that generates the header file as a dependency. At any rate, it can be solved in a follow-up PR.